prov/rxm/src/rxm_ep.c

```diff
@@ -799,6 +799,131 @@ rxm_ep_format_tx_res(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn,
 	return ret;
 }
 
+void rxm_conn_handle_postponed_tx_op(struct rxm_ep *rxm_ep,
+				     struct util_cmap_handle *handle)
+{
+	struct rxm_tx_entry *tx_entry;
+	struct rxm_tx_buf *tx_buf;
+	const size_t pkt_size = rxm_pkt_size;
+	ssize_t ret;
+	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
+
+	while (!dlist_empty(&rxm_conn->postponed_tx_list)) {
+		size_t tx_size;
+
+		dlist_pop_front(&rxm_conn->postponed_tx_list, struct rxm_tx_entry,
+				tx_entry, postponed_entry);
+
+		tx_size = pkt_size + tx_entry->tx_buf->pkt.hdr.size;
+		tx_buf = tx_entry->tx_buf;
+		tx_buf->pkt.ctrl_hdr.conn_id = handle->remote_key;
+
+		if ((tx_size <= rxm_ep->msg_info->tx_attr->inject_size) &&
+		    (tx_entry->flags & FI_INJECT) && !(tx_entry->flags & FI_COMPLETION))  {
+			ret = fi_inject(rxm_conn->msg_ep, &tx_buf->pkt, pkt_size, 0);
+			if (OFI_UNLIKELY(ret)) {
+				FI_DBG(&rxm_prov, FI_LOG_EP_DATA,
+				       "fi_inject for MSG provider failed\n");
+				rxm_cntr_incerr(rxm_ep->util_ep.tx_cntr);
+			} else {
+				rxm_cntr_inc(rxm_ep->util_ep.tx_cntr);
+			}
+			/* Release resources for futher reuse */
+			goto cur_fn;
+		} else if (tx_entry->tx_buf->pkt.hdr.size >
+				rxm_ep->rxm_info->tx_attr->inject_size) {
+			struct rxm_rma_iov *rma_iov =
+				(struct rxm_rma_iov *)&tx_entry->tx_buf->pkt.data;
+			ret = sizeof(*rma_iov) + sizeof(*rma_iov->iov) * tx_entry->count;
+			if (pkt_size + ret <= rxm_ep->msg_info->tx_attr->inject_size) {
+				RXM_LOG_STATE(FI_LOG_EP_DATA, tx_entry->tx_buf->pkt, RXM_TX, RXM_LMT_TX);
+				ret = fi_inject(rxm_conn->msg_ep, &tx_buf->pkt, pkt_size + ret, 0);
+				if (OFI_UNLIKELY(ret)) {
+					FI_DBG(&rxm_prov, FI_LOG_EP_DATA,
+					       "fi_inject for MSG provider failed\n");
+					goto cur_fn;
+				}
+
+				RXM_LOG_STATE(FI_LOG_CQ, tx_entry->tx_buf->pkt, RXM_LMT_TX, RXM_LMT_ACK_WAIT);
+				tx_entry->state = RXM_LMT_ACK_WAIT;
+			} else {
+				ret = fi_send(rxm_conn->msg_ep, &tx_buf->pkt, pkt_size + ret,
+					      tx_buf->hdr.desc, 0, tx_entry);
+				if (OFI_UNLIKELY(ret)) {
+					if (ret == -FI_EAGAIN)
+						rxm_ep_progress_multi(&rxm_ep->util_ep);
+					else
+						FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
+							"fi_send for MSG provider failed\n");
+					goto cur_fn;
+				}
+				RXM_LOG_STATE(FI_LOG_EP_DATA, tx_entry->tx_buf->pkt, RXM_TX, RXM_LMT_TX);
+				tx_entry->state = RXM_LMT_TX;
+			}
+		} else {
+			ret = fi_send(rxm_conn->msg_ep, &tx_buf->pkt, tx_size,
+				      tx_buf->hdr.desc, 0, tx_entry);
+			if (OFI_UNLIKELY(ret)) {
+				if (ret == -FI_EAGAIN)
+					rxm_ep_progress_multi(&rxm_ep->util_ep);
+				else
+					FI_WARN(&rxm_prov, FI_LOG_EP_DATA,
+						"fi_send for MSG provider failed\n");
+				goto cur_fn;
+			}
+		}
+		continue;
+cur_fn:
+	rxm_tx_buf_release(rxm_ep, tx_entry->tx_buf);
+	rxm_tx_entry_release(&rxm_ep->send_queue, tx_entry);
+	}
+}
```

Comment:
This is a fairly long function.  Do we need to have so many optimizations when handling posting deferred transmits, or can we get by with fewer more general cases (e.g. just call sendmsg() from a loop)?  It seems that there should be some sort of code sharing between processing the postponed list and the regular transmit paths.

Also, to support this in general, we need to be able to queue any type of transfer, including RMA.  How is that handled?

Reply:
Thanks, the RMA transfer must be handled in the same way.

Reply:
> or can we get by with fewer more general cases (e.g. just call sendmsg() from a loop)?

Unfortunately, there are some dependencies (please, review updates) that can't be handled by calling fi_sendmsg() of core provider from a loop.